graph/run-dynamic.go

```diff
@@ -263,48 +263,59 @@ func ConvergeDynamicThread[EP EPP[E], V VPI[V], E EPI[E], M MVI[M], N any, A Alg
 				// There is a request for synchronously syncing topology (blocking), acknowledge here as we have no more topology events to process.
 				bspSync = false
 				gt.Response <- ACK
-			} else if remitCount == 0 {
-				// Want topology events, but none found -- we should relax.
-				topFail++
-				gt.Status = BACKOFF_TOP
-				utils.BackOff(topFail)
-				if timeStates && tidx == 0 {
-					curr = time.Now()
-					gt.LoopTimes[5] += curr.Sub(prev)
-					prev = curr
-				}
-			} else {
-				topFail = 0
+			} else if remitCount == 0 && !epoch {
+				topFailed = true
 			}
 
 			if strucClosed {
 				doneEvents <- struct{}{}
-			} else if TOPOLOGY_FIRST || blockAlgIfTop {
-				continue
 			}
 		}
+		if topFailed {
+			topFail++
+		} else {
+			topFail = 0
+		}
+
+		algFailed := false
+		if !(!strucClosed && (TOPOLOGY_FIRST || blockAlgIfTop)) {
+			// Process algorithm messages. Check for algorithm termination if needed.
+			gt.Status = APPLY_MSG
+			checkTerm := (strucClosed && remitClosed) || (epoch && remitCount == 0 && topCount == 0)
+			completed, algCount = ProcessMessages[V, E, M, N](alg, g, gt, checkTerm)
+			if completed && epoch {
+				//log.Debug().Msg("T[" + utils.F("%02d", tidx) + "] completed epoch")
+				gt.Status = DONE
+				gt.Response <- ACK
 
-		// Process algorithm messages. Check for algorithm termination if needed.
-		gt.Status = APPLY_MSG
-		checkTerm := (strucClosed && remitClosed) || (epoch && remitCount == 0 && count == 0)
-		completed, count = ProcessMessages[V, E, M, N](alg, g, gt, checkTerm)
-		if completed && epoch {
-			//log.Debug().Msg("T[" + utils.F("%02d", tidx) + "] completed epoch")
-			gt.Status = DONE
-			gt.Response <- ACK
-
-			resp := <-gt.Command // BLOCK and wait for resume
-			if resp != RESUME {
-				log.Panic().Msg("Expected to resume after blocked")
+				resp := <-gt.Command // BLOCK and wait for resume
+				if resp != RESUME {
+					log.Panic().Msg("Expected to resume after blocked")
+				}
+				epoch = false
+				completed = false
+			} else if algCount == 0 {
+				algFailed = true
 			}
-			epoch = false
-			completed = false
-			g.TerminateVotes[tidx] = 0
 		}
-
-		if count == 0 {
+		if algFailed {
 			algFail++
-			if algFail%10 == 0 {
+		} else {
+			algFail = 0
+		}
+
+		// Backoff only when there are no topological events and algorithmic event
```

Comment:
This will break with higher query granularity (backs off too quick; because it is incremented even when that increment isn't used for a call to back off).
You would need to move the increments down here